### PR TITLE
Show URL on inactive tab in split view

### PR DESCRIPTION
**Pull request by sangwoo108**
For security reason, we should let users know where they are.

Resolves https://github.com/brave/brave-browser/issues/39152

image

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Comment by diracdeltas** [comment]
is there a m1 build to test this out? if not i'll try the regular macos build when i get back to my non-M1 mac

**Comment by sangwoo108** [comment]
> is there a m1 build to test this out? if not i'll try the regular macos build when i get back to my non-M1 mac

Sorry, I couldn't find one 😞

File: browser/ui/views/frame/brave_browser_view.h
Comment (fallaciousreasoning) [inline browser/ui/views/frame/brave_browser_view.h:22]: nit: can we forward declare this? I'm always a bit hazy about if you can do it with `raw_ptr`

File: browser/ui/views/split_view/split_view_location_bar.cc
Comment (fallaciousreasoning) [inline browser/ui/views/split_view/split_view_location_bar.cc:43]: ~~I wonder if its worth using the (overridden) Chromium icon here instead, so this always stays in sync with the Omnibox~~ nvm - I checked and it uses the filled icon so we can't reuse it.

:+1:

File: browser/ui/views/split_view/split_view_location_bar.cc
Comment (fallaciousreasoning) [inline browser/ui/views/split_view/split_view_location_bar.cc:160]: It might be worth using on of the Chromium methods for formatting this:
- `AppBrowserController::FormatUrlOrigin` does something like what we (probably) want here
- Otherwise there's `url_formatter::FormatUrl`

We should probably have the security team look at this - formatting URLs can get pretty complicated - I think Chromium has some logic somewhere for handling lookalike characters (i.e. from cyrillic) and converting unicode to punycode

File: browser/ui/views/split_view/split_view_location_bar.cc
Comment (fallaciousreasoning) [inline browser/ui/views/split_view/split_view_location_bar.cc:159]: I think other places use `GetVisibleURL` rather than last committed so that the correct URL shows ASAP

File: browser/ui/views/split_view/split_view_location_bar.cc
Comment (fallaciousreasoning) [reply browser/ui/views/split_view/split_view_location_bar.cc:165]: I'm pretty sure @thypon is working to remove this false positive

File: browser/ui/views/split_view/split_view_location_bar.cc
Comment (fallaciousreasoning) [inline browser/ui/views/split_view/split_view_location_bar.cc:180]: Can we use the `SecurityLevel` of the WebContents instead of doing this?

```
SecurityStateTabHelper::FromWebContents(contents)->GetSecurityLevel() != SECURE
```

there are a lot of edge cases here (unicode, SSL status, certificates, insecure subresources, reader mode) and we probably don't want to worry about them all.

Probably a good idea to rope security in for this too

File: browser/ui/views/split_view/split_view_location_bar.cc
Comment (fallaciousreasoning) [inline browser/ui/views/split_view/split_view_location_bar.cc:185]: This looks really good!

File: browser/ui/views/split_view/split_view_location_bar.h
Comment (fallaciousreasoning) [inline browser/ui/views/split_view/split_view_location_bar.h:43]: Might want to listen to `DidChangeVisibleSecurityState` too

**Review by fallaciousreasoning — COMMENTED** [review]
Looks great! Sorry for all the comments about security - I just remember there being heaps of edge cases when I was working on the `CustomTabBarView` :laughing:

File: browser/ui/views/frame/brave_browser_view.h
Comment (sangwoo108) [reply browser/ui/views/frame/brave_browser_view.h]: Oh sure. My autocomplete filled it.

File: browser/ui/views/split_view/split_view_location_bar.cc
Comment (sangwoo108) [reply browser/ui/views/split_view/split_view_location_bar.cc]: Good call! That'd be much better. Left only about scheme and chrome scheme, as they are considered insecure by the API.

File: browser/ui/views/split_view/split_view_location_bar.cc
Comment (sangwoo108) [reply browser/ui/views/split_view/split_view_location_bar.cc:212]: Thanks 😊

File: browser/ui/views/split_view/split_view_location_bar.cc
Comment (sangwoo108) [reply browser/ui/views/split_view/split_view_location_bar.cc]: Nice! But hope security team can double check this, as I was sometimes asked to use committed url.

File: browser/ui/views/split_view/split_view_location_bar.cc
Comment (sangwoo108) [reply browser/ui/views/split_view/split_view_location_bar.cc]: Also great suggestion. Done.

File: browser/ui/views/frame/brave_browser_view.h
Comment (fallaciousreasoning) [reply browser/ui/views/frame/brave_browser_view.h]: Mine does it all the time! It's super helpful mostly :laughing:

File: browser/ui/views/split_view/split_view_location_bar.cc
Comment (fallaciousreasoning) [reply browser/ui/views/split_view/split_view_location_bar.cc]: Yup definitely wise :laughing: - I think the reasoning for using `GetVisibleURL` would be that its a visible URL :laughing: (I think its what the Omnibox/CustomTabBar display) but lets defer to the experts :closed_lock_with_key:

**Review by fallaciousreasoning — APPROVED** [review]
Sweet, LGTM @sangwoo108. Let's see what the security folks reckon :smile:

File: browser/ui/views/split_view/split_view_location_bar.cc
Comment (diracdeltas) [inline browser/ui/views/split_view/split_view_location_bar.cc:174]: seems safer to use [GetLastCommittedURL](https://source.chromium.org/chromium/chromium/src/+/main:content/public/browser/web_contents.h;bpv=1;bpt=1;l=430?q=getvisibleurl&ss=chromium&gsn=GetLastCommittedURL&gs=KYTHE%3A%2F%2FKswCCqMBa3l0aGU6Ly9jaHJvbWl1bS5nb29nbGVzb3VyY2UuY29tL2NvZGVzZWFyY2gvY2hyb21pdW0vc3JjLy9tYWluP2xhbmc9YyUyQiUyQj9wYXRoPWNvbnRlbnQvcHVibGljL2Jyb3dzZXIvd2ViX2NvbnRlbnRzLmgjVllEVHd1UWY5QmY5SW8zdk1NWVViYzExRU0xYVY0TU5zNlVZOFBoc0FUOAqjAWt5dGhlOi8vY2hyb21pdW0uZ29vZ2xlc291cmNlLmNvbS9jb2Rlc2VhcmNoL2Nocm9taXVtL3NyYy8vbWFpbj9sYW5nPWMlMkIlMkI_cGF0aD1jb250ZW50L3B1YmxpYy9icm93c2VyL3dlYl9jb250ZW50cy5oI2ZfVEhXLUQ2bm55aVdLaUd0TGIzVUwwazctLUU5Nmh0SUhRQTlEeTdQVUE%3D)

File: browser/ui/views/split_view/split_view_location_bar.cc
Comment (diracdeltas) [inline browser/ui/views/split_view/split_view_location_bar.cc:176]: on second thought, why not just show the full origin here? i'm not sure what happens if the url is `blob:` or `view-source:` but it seems important to include the scheme

File: browser/ui/views/split_view/split_view_location_bar.cc
Comment (diracdeltas) [inline browser/ui/views/split_view/split_view_location_bar.cc:206]: seems safer to use [GetLastCommittedURL](https://source.chromium.org/chromium/chromium/src/+/main:content/public/browser/web_contents.h;bpv=1;bpt=1;l=430?q=getvisibleurl&ss=chromium&gsn=GetLastCommittedURL&gs=KYTHE%3A%2F%2FKswCCqMBa3l0aGU6Ly9jaHJvbWl1bS5nb29nbGVzb3VyY2UuY29tL2NvZGVzZWFyY2gvY2hyb21pdW0vc3JjLy9tYWluP2xhbmc9YyUyQiUyQj9wYXRoPWNvbnRlbnQvcHVibGljL2Jyb3dzZXIvd2ViX2NvbnRlbnRzLmgjVllEVHd1UWY5QmY5SW8zdk1NWVViYzExRU0xYVY0TU5zNlVZOFBoc0FUOAqjAWt5dGhlOi8vY2hyb21pdW0uZ29vZ2xlc291cmNlLmNvbS9jb2Rlc2VhcmNoL2Nocm9taXVtL3NyYy8vbWFpbj9sYW5nPWMlMkIlMkI_cGF0aD1jb250ZW50L3B1YmxpYy9icm93c2VyL3dlYl9jb250ZW50cy5oI2ZfVEhXLUQ2bm55aVdLaUd0TGIzVUwwazctLUU5Nmh0SUhRQTlEeTdQVUE%3D)

File: browser/ui/views/split_view/split_view_location_bar.cc
Comment (diracdeltas) [inline browser/ui/views/split_view/split_view_location_bar.cc:202]: what happens if you load https://expired.badssl.com/ (both before and after clicking through the certificate error)?

File: browser/ui/views/split_view/split_view_location_bar.cc
Comment (diracdeltas) [inline browser/ui/views/split_view/split_view_location_bar.cc:208]: we probably have to include a bunch more protocols here, like blob, data, file, etc. not sure if there's a better way to do it because it's hard to enumerate all of them.

i guess though if we are showing the full origin, including scheme, as i suggested above, there's no need to show any security indicator unless it's an invalid HTTPS page (in which case we can put a slash through the HTTPS as is done in the url bar)

File: browser/ui/views/split_view/split_view_location_bar.cc
Comment (sangwoo108) [reply browser/ui/views/split_view/split_view_location_bar.cc:202]: image

image

Comment (diracdeltas) [reply browser/ui/views/split_view/split_view_location_bar.cc:202]: thanks

File: browser/ui/views/split_view/split_view_location_bar.cc
Comment (diracdeltas) [reply browser/ui/views/split_view/split_view_location_bar.cc]: i think this should always reflect the location of the current page in the webcontents, so GetLastCommittedURL was right

File: browser/ui/views/split_view/split_view_location_bar.cc
Comment (sangwoo108) [reply browser/ui/views/split_view/split_view_location_bar.cc:176]: Sure. It'll be like

image

image

image

I also couldn't find nice way to iterate "safe" schemes. Do you think we can omit the icon then?

Comment (sangwoo108) [reply browser/ui/views/split_view/split_view_location_bar.cc:176]: I found that work for us in that case

```
  auto security_level = SecurityStateTabHelper::FromWebContents(web_contents())
                            ->GetSecurityLevel();
  return security_level == security_state::SecurityLevel::SECURE ||
         security_level == security_state::SecurityLevel::NONE;
```

Comment (diracdeltas) [reply browser/ui/views/split_view/split_view_location_bar.cc:176]: IMO it is more informative to omit the icon and just show the full origin (preferably with a slash through HTTPS on invalid certs)

**Comment by sangwoo108** [comment]
In case of long host name, as mentioned in the security review, it'll be truncated like

image

File: browser/ui/views/split_view/split_view_location_bar.cc
Comment (thypon) [reply browser/ui/views/split_view/split_view_location_bar.cc:176]: Probably better to show the scheme, only if it's not `https` here

File: browser/ui/views/split_view/split_view_location_bar.cc
Comment (thypon) [reply browser/ui/views/split_view/split_view_location_bar.cc:186]: removed in fact

File: browser/ui/views/split_view/split_view_location_bar.cc
Comment (sangwoo108) [reply browser/ui/views/split_view/split_view_location_bar.cc:176]: So, do we want it to be like
* Hide scheme when the security level is secure - https:// and valid cert -> example.com
* Show slashed when it's https:// but not valid - ~~https~~://example.com
* Show scheme when the security level is  none(neutral, like chrome:// view-source) - chrome://newtab

And remove the icon entirely. Am I correct?

Comment (diracdeltas) [reply browser/ui/views/split_view/split_view_location_bar.cc:176]: IMO it'd be confusing to only hide scheme if HTTPS. i understand it takes up more space though.

otherwise sgtm

Comment (thypon) [reply browser/ui/views/split_view/split_view_location_bar.cc:176]: @diracdeltas it's the same behavior of the omnibar. I feel being not consistent is confusing.
Also it looks bad visually, but that's secondary.

Comment (thypon) [reply browser/ui/views/split_view/split_view_location_bar.cc:176]: I would leave the icon for not secure (none // not secure, and when we strikethrough the ~~https://~~

Comment (thypon) [reply browser/ui/views/split_view/split_view_location_bar.cc:176]: As yan mentioned, there is a config to show the full URL (with the schema), all the time. We should probably follow that configuration and show `https` all the times, in that particular case.

I don't see much point in having drift from the omnibar behaviour

Comment (diracdeltas) [reply browser/ui/views/split_view/split_view_location_bar.cc:176]: it turns out i have the `Always show full URLs` setting enabled whereas @thypon does not. in any case, we agree it'd be least confusing if this mini-URL view just showed the same as whatever the user would see in the omnibox minus the URL path (so following that setting)

Comment (diracdeltas) [reply browser/ui/views/split_view/split_view_location_bar.cc:176]: yes, also fine with me to show the icon as long as it matches what is shown in the omnibox 100% of the time. whatever chromium does to determin the omnibox icon, we should just copy that logic.

File: browser/ui/views/split_view/split_view_location_bar.cc
Comment (fallaciousreasoning) [reply browser/ui/views/split_view/split_view_location_bar.cc]: I think we probably want `GetVisibleURL` here @diracdeltas - its what the `LocationBarModelDelegate` and the `CustomTabBar` are using (or equivalent).

The `CustomTabBar` goes directly through `GetVisibleURL` while the `ChromeLocationBarModelDelegate` does the following:

```cpp
content::NavigationEntry* ChromeLocationBarModelDelegate::GetNavigationEntry()
    const {
  content::NavigationController* controller = GetNavigationController();
  return controller ? controller->GetVisibleEntry() : nullptr;
}

bool ChromeLocationBarModelDelegate::GetURL(GURL* url) const {
  DCHECK(url);
  content::NavigationEntry* entry = GetNavigationEntry();
  if (!entry || entry->IsInitialEntry())
    return false;

  *url = entry->GetVirtualURL();
  return true;
}
```

for reference, this is the implementation of `GetVisibleURL` in `WebContentsImpl`

```cpp
const GURL& WebContentsImpl::GetVisibleURL() {
  // We may not have a navigation entry yet.
  NavigationEntry* entry = GetController().GetVisibleEntry();
  return entry ? entry->GetVirtualURL() : GURL::EmptyGURL();
}
```

and this is the comment on `GetVisibleURL` in `web_contents.h`

```cpp
  // Gets the virtual URL currently being displayed in the URL bar, if there is
  // one. This URL might be a pending navigation that hasn't committed yet, so
  // it is not guaranteed to match the current page in this WebContents.
  virtual const GURL& GetVisibleURL() = 0;
```

I think this makes sense - if you're on `https://mybank.com` and you somehow navigate to `https://evil-phising-mybank.com` you don't want to be showing `https://mybank.com` in the URL bar while the phising site is loading.

**Comment by fallaciousreasoning** [comment]
sorry for all the comments @sangwoo108 but I think in other places when Chromium truncates the origin they truncate the start of the origin:

https://long-subdomain.mybank.com.evil-phising.com

should truncate to

`https://...evil-phising.com`

not

`https://long-subdomain.mybank.com...`

I'll try and find the logic Chromium uses

**Comment by sangwoo108** [comment]
> sorry for all the comments @sangwoo108 @diracdeltas but I think in other places when Chromium truncates the origin they truncate the start of the origin:

Nice find @fallaciousreasoning ! I thought so, but at least omnibox truncates the url from the tail. @diracdeltas , what's your take on this?

File: browser/ui/views/split_view/split_view_location_bar.cc
Comment (diracdeltas) [reply browser/ui/views/split_view/split_view_location_bar.cc]: @bridiver do you have thoughts here? 

my reasoning is that this mini/readonly URL display should reflect what page is currently loaded, whereas the full URL bar (because the user may enter URLs into it) has to show pending navigations.

that is if you're on google.com in a split view, and the user initiates a navigation to othersite.com which takes a long time to load, it makes sense to me that the urlbar would show othersite.com but i'd still expect the split view's readonly URL display to show google.com. however i could go either way on this.

note that there have been historically a bunch of URL spoofing issues like https://issues.chromium.org/issues/40096119 with the omnibox

**Comment by diracdeltas** [comment]
> Nice find @fallaciousreasoning ! I thought so, but at least omnibox truncates the url from the tail. @diracdeltas , what's your take on this?

security-wise it actually is better to truncate from the left so if chromium has a builtin way to do this let's use it. note that it gets tricky with languages like arabic that are read from right-to-left instead of left-to-right; then the truncation has to happen on the right.

File: browser/ui/views/split_view/split_view_location_bar.cc
Comment (sangwoo108) [reply browser/ui/views/split_view/split_view_location_bar.cc]: > my reasoning is that this mini/readonly URL display should reflect what page is currently loaded, whereas the full URL bar (because the user may enter URLs into it) has to show pending navigations.

I agree with this. Unlike the omnibox, the mini URL bar is for the inactive tab, so we should prioritize what's actually loaded over what the user typed or what's still loading(pending).

File: browser/ui/views/split_view/split_view_location_bar.cc
Comment (sangwoo108) [reply browser/ui/views/split_view/split_view_location_bar.cc:176]: Thanks all, I'll apply the logic above only when "Always show full URLs" is off. And leave icon as is.

File: browser/ui/views/split_view/split_view_location_bar.cc
Comment (fallaciousreasoning) [reply browser/ui/views/split_view/split_view_location_bar.cc]: okay, sgtm :smile:

File: browser/ui/views/split_view/split_view_location_bar.cc
Comment (fallaciousreasoning) [reply browser/ui/views/split_view/split_view_location_bar.cc]: Just as a quick sanity check - is it easier to check insecure schemes? (http, https ect)?

Comment (sangwoo108) [reply browser/ui/views/split_view/split_view_location_bar.cc]: Yeah, The latest version is checking like

```
  auto security_level = SecurityStateTabHelper::FromWebContents(web_contents())
                            ->GetSecurityLevel();
  return security_level == security_state::SecurityLevel::SECURE ||
         security_level == security_state::SecurityLevel::NONE;
```

The NONE type would include them, so we can filter only the insecure state.

**Comment by sangwoo108** [comment]
Hello, @diracdeltas @thypon @fallaciousreasoning . I think I addressed we discussed. Could you take another look? This is how it would be with "Always Show Full URLs" off.

### valid https
image

### With invalid cert
image

### security level == none, such as chrome:// scheme or view-source: scheme
image
image

### mere http
image

### long host - truncates from the left
image

**Review by fallaciousreasoning — APPROVED** [review]
LGTM. Added a screenshot of one more state (bad ssl, long domain truncation), which looks great:

image

Nice work @sangwoo108, heaps of weird cases on this one!

**Review by diracdeltas — APPROVED** [review]
i can't get the .dmg from CI to run on any of my macbooks (both M1 and intel), so i haven't tried it out, but the logic seems fine to me

**Comment by sangwoo108** [comment]
@thypon , Do you have any other comments?

**Review by thypon — APPROVED** [review]

File: browser/ui/views/split_view/split_view_location_bar.cc
Comment (bridiver) [inline browser/ui/views/split_view/split_view_location_bar.cc:222]: This is definitely not ok. We should not be trying to duplicate the logic for determining the current url here and GetLastCommittedURL is not the right method to use. This entire class is extremely problematic and this location is going to be subject to spoofing because it isn't using the correct logic.

Comment (bridiver) [reply browser/ui/views/split_view/split_view_location_bar.cc:222]: `GetVisibleURL` is the correct method for the url display, but we should not be duplicating the logic of determining the url and there are already existing methods for formatting, etc...

File: browser/ui/views/split_view/split_view_location_bar.cc
Comment (bridiver) [inline browser/ui/views/split_view/split_view_location_bar.cc:244]: there is a location bar model delegate to handle this stuff, like the actual url, we should not be writing our own logic for this stuff. This entire class is a major problem and I think we should revert this PR because we need a completely different approach here.

**Comment by sangwoo108** [comment]
Let me create a revert. Thanks.

**Comment by sangwoo108** [comment]
Created : https://github.com/brave/brave-core/pull/24408

**Comment by ShivanKaul** [comment]
@diracdeltas and @thypon captured all my concerns here, thanks

File: browser/ui/views/split_view/split_view_location_bar.cc
Comment (diracdeltas) [reply browser/ui/views/split_view/split_view_location_bar.cc:222]: there was some discussion here: https://github.com/brave/brave-core/pull/24293#discussion_r1650211857. i thought GetLastCommittedURL was correct in this case because the intent of this readonly display is to show what's currently loaded, not what is about to be loaded

Comment (diracdeltas) [reply browser/ui/views/split_view/split_view_location_bar.cc:222]: we discussed over DM and i now agree with brian

**Pull request merged**
